src/experimental/patterns/ec2-app.ts

```diff
@@ -0,0 +1,282 @@
+import type { IAspect, Stack } from "aws-cdk-lib";
+import { Aspects, CfnParameter, Duration } from "aws-cdk-lib";
+import type { CfnAutoScalingGroup } from "aws-cdk-lib/aws-autoscaling";
+import { CfnScalingPolicy, ScalingProcess, UpdatePolicy } from "aws-cdk-lib/aws-autoscaling";
+import { Effect, Policy, PolicyStatement } from "aws-cdk-lib/aws-iam";
+import type { IConstruct } from "constructs";
+import { GuAutoScalingGroup } from "../../constructs/autoscaling";
+import { GuStack } from "../../constructs/core";
+import type { GuEc2AppProps } from "../../patterns";
+import { GuEc2App } from "../../patterns";
+import { isSingletonPresentInStack } from "../../utils/singleton";
+
+/**
+ * An `Aspect` that adjusts the properties of an AutoScaling Group using an `AutoScalingRollingUpdate` update policy.
+ *
+ * It'll unset the `DesiredCapacity` of the ASG as a rolling update sets `Desired` back to the template version,
+ * undoing any changes that a scale-out event may have done.
+ * Having `DesiredCapacity` unset ensures the service remains at-capacity at all times.
+ *
+ * It'll also make the `MinInstancesInService` property dynamic via a CFN Parameter that Riff-Raff will set.
+ * The value depends on the current capacity of the ASG:
+ *  - If the service is running normally, it'll be set to the `MinSize` capacity
+ *  - If the service is partially scaled, it'll be set to the current `DesiredCapacity`
+ *  - If the service is fully scaled, it'll be set to (at least) `MaxSize` - 1
+ *
+ * @privateRemarks
+ * - Temporarily implemented as a singleton to ensure only one instance is added to a stack,
+ *   else multiple attempts to add the same CFN Parameter will be made and fail.
+ * - Once out of experimental, instantiate this `Aspect` directly in {@link GuStack}.
+ *
+ * @see https://github.com/guardian/testing-asg-rolling-update
+ */
+class HorizontallyScalingDeploymentProperties implements IAspect {
```

Comment:
This is implemented as an [`Aspect`](https://docs.aws.amazon.com/cdk/v2/guide/aspects.html) to handle cases when a scaling policy is added after the call to `super` of an `GuStack` implementer's constructor.